File: antd-dart/test/finalize_with_retry_test.dart
import 'dart:convert';

import 'package:antd_client/antd_client.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:test/test.dart';

import '../example/finalize_with_retry.dart';

/// Leaves `retryable` out of the body, as an antd < 0.14.0 daemon does.
const _absent = Object();

/// The daemon's structured `PARTIAL_UPLOAD` 502 body for a 312-chunk upload
/// with [failed] chunks still unstored. [retryable] is written as given (any
/// JSON value, null included), or left out when it is [_absent].
http.Response _partial(int failed, {Object? retryable = true}) {
  final stored = 312 - failed;
  return http.Response(
    jsonEncode({
      'error': 'Partial upload: $stored/312 chunks stored, $failed failed '
          'after retries',
      'code': 'PARTIAL_UPLOAD',
      'chunks_stored': stored,
      'chunks_failed': failed,
      'total_chunks': 312,
      if (!identical(retryable, _absent)) 'retryable': retryable,
    }),
    502,
    headers: {'content-type': 'application/json'},
  );
}

http.Response _finalized() => http.Response(
      jsonEncode({
        'data_map': 'dm',
        'data_map_address': 'dm-addr',
        'chunks_stored': 312,
      }),
      200,
      headers: {'content-type': 'application/json'},
    );

http.Response _error(int status, String message) => http.Response(
      jsonEncode({'error': message}),
      status,
      headers: {'content-type': 'application/json'},
    );

/// Answers `POST /v1/upload/finalize` from [responses] in order, repeating
/// the last one, and counts the finalize calls. Errors go through the real
/// REST error mapping in [AntdClient].
class _FinalizeDaemon {
  _FinalizeDaemon(this.responses);

  final List<http.Response Function()> responses;
  final backoffs = <int>[];
  final logs = <String>[];
  var calls = 0;

  Future<FinalizeUploadResult> run({int maxAttempts = 5}) {
    final client = AntdClient(httpClient: MockClient((request) async {
      expect('${request.method} ${request.url.path}',
          equals('POST /v1/upload/finalize'));
      expect(jsonDecode(request.body)['upload_id'], equals('up-1'));
      final i = calls < responses.length ? calls : responses.length - 1;
      calls++;
      return responses[i]();
    }));
    addTearDown(client.close);
    return finalizeWithRetry(
      client,
      'up-1',
      {'0xq': '0xt'},
      maxAttempts: maxAttempts,
      backoff: (attempt) {
        backoffs.add(attempt);
        return Duration.zero;
      },
      log: logs.add,
    );
  }
}

/// Runs [daemon] and returns what it threw, with the stack trace.
Future<(Object, StackTrace)> _thrown(_FinalizeDaemon daemon,
    {int maxAttempts = 5}) async {
  try {
    await daemon.run(maxAttempts: maxAttempts);
  } catch (e, st) {
    return (e, st);
  }
  throw TestFailure('finalizeWithRetry returned instead of throwing');
}

/// Throws, from every finalizeUpload call, the [PartialUploadError] the gRPC
/// client builds from an ABORTED status [message], and counts the calls. The
/// SDK's gRPC parser, not a hand-set flag, decides what the helper sees.
class _GrpcMessageClient extends AntdClient {
  _GrpcMessageClient(this.message)
      : super(
            httpClient: MockClient((_) async =>
                throw StateError('finalizeUpload is overridden')));

  final String message;
  var calls = 0;

  @override
  Future<FinalizeUploadResult> finalizeUpload(
    String uploadId,
    Map<String, String> txHashes,
  ) async {
    calls++;
    throw PartialUploadError.fromMessage(message);
  }
}

/// Counts prefix of a gRPC PARTIAL_UPLOAD message for a 312-chunk upload
/// with 12 chunks still unstored.
const _grpcCounts =
    'Partial upload: 300/312 chunks stored, 12 failed after retries: quorum';

Matcher _partialError({
  required int failed,
  required bool retryable,
  bool retentionKnown = true,
}) =>
    isA<PartialUploadError>()
        .having((e) => e.statusCode, 'statusCode', 502)
        .having((e) => e.chunksStored, 'chunksStored', 312 - failed)
        .having((e) => e.chunksFailed, 'chunksFailed', failed)
        .having((e) => e.totalChunks, 'totalChunks', 312)
        .having((e) => e.retryable, 'retryable', retryable)
        .having((e) => e.retentionKnown, 'retentionKnown', retentionKnown);

void main() {
  group('finalizeWithRetry', () {
    test('returns the result when a retry stores the remainder', () async {
      final daemon = _FinalizeDaemon([() => _partial(12), _finalized]);
      final result = await daemon.run();
      expect(result.chunksStored, equals(312));
      expect(result.dataMapAddress, equals('dm-addr'));
      expect(daemon.calls, equals(2));
      expect(daemon.backoffs, equals([1]));
    });

    test('a first-call success makes one call', () async {
      final daemon = _FinalizeDaemon([_finalized]);
      expect((await daemon.run()).dataMap, equals('dm'));
      expect(daemon.calls, equals(1));
      expect(daemon.backoffs, isEmpty);
    });

    test('attempt exhaustion rethrows the last PartialUploadError unchanged',
        () async {
      // chunksFailed keeps shrinking, so only maxAttempts stops the loop.
      final daemon = _FinalizeDaemon(
          [() => _partial(12), () => _partial(8), () => _partial(4)]);
      final (error, stack) = await _thrown(daemon, maxAttempts: 3);
      expect(error, _partialError(failed: 4, retryable: true));
      // Catchable by `on PartialUploadError`, `on NetworkError` and
      // `on Exception` alike; a StateError would escape all three.
      expect(error, isA<NetworkError>());
      expect(error, isA<Exception>());
      expect(error, isNot(isA<Error>()));
      // Rethrown, not rewrapped: the trace still starts in the client's
      // error mapping rather than in the retry helper.
      expect(stack.toString(), contains('package:antd_client/src/client.dart'));
      expect(daemon.calls, equals(3));
      expect(daemon.backoffs, equals([1, 2]));
    });

    test('maxAttempts: 1 makes one call and rethrows the partial error',
        () async {
      final daemon = _FinalizeDaemon([() => _partial(12)]);
      final (error, _) = await _thrown(daemon, maxAttempts: 1);
      expect(error, _partialError(failed: 12, retryable: true));
      expect(error, isA<Exception>());
      expect(daemon.calls, equals(1));
      expect(daemon.backoffs, isEmpty);
    });

    test('an unchanged chunksFailed stops early and rethrows', () async {
      final daemon =
          _FinalizeDaemon([() => _partial(12), () => _partial(12), _finalized]);
      final (error, _) = await _thrown(daemon);
      expect(error, _partialError(failed: 12, retryable: true));
      expect(error, isA<Exception>());
      expect(daemon.calls, equals(2));
      expect(daemon.backoffs, equals([1]));
    });

    test('a growing chunksFailed stops early and rethrows', () async {
      final daemon = _FinalizeDaemon(
          [() => _partial(8), () => _partial(4), () => _partial(6)]);
      final (error, _) = await _thrown(daemon);
      expect(error, _partialError(failed: 6, retryable: true));
      expect(daemon.calls, equals(3));
      expect(daemon.backoffs, equals([1, 2]));
    });

    test('nothing retained (retryable false) is rethrown after one call',
        () async {
      // The daemon confirmed it kept nothing: the caller re-prepares.
      final daemon = _FinalizeDaemon([() => _partial(12, retryable: false)]);
      final (error, _) = await _thrown(daemon);
      expect(error, _partialError(failed: 12, retryable: false));
      expect(daemon.calls, equals(1));
      expect(daemon.backoffs, isEmpty);
    });

    test('retention unknown stops at once and rethrows the typed error',
        () async {
      // No boolean retryable (a daemon before 0.14.0, or a mistyped flag):
      // the daemon may still hold the paid attempt, so the helper makes one
      // finalize call and hands the typed error back without retrying,
      // re-preparing or paying.
      for (final retryable in [_absent, null, 'true', 1]) {
        final label = identical(retryable, _absent) ? 'absent' : '$retryable';
        final daemon =
            _FinalizeDaemon([() => _partial(12, retryable: retryable)]);
        final (error, _) = await _thrown(daemon);
        expect(error,
            _partialError(failed: 12, retryable: false, retentionKnown: false),
            reason: label);
        expect(error, isA<Exception>(), reason: label);
        expect(daemon.calls, equals(1), reason: label);
        expect(daemon.backoffs, isEmpty, reason: label);
        expect(daemon.logs.single, contains('retention is unknown'),
            reason: label);
      }
    });

    // End to end from the gRPC status text: whenever the daemon's closing
    // retention hint cannot be read, the helper must stop at once with the
    // reconcile advice, never retry, re-prepare or pay.
    const unreadable = {
      'no hint': _grpcCounts,
      // The review's reproducer: the retained hint cut short.
      'a truncated retained hint': '$_grpcCounts (paid attempt retai',
      'an unclosed retained hint':
          '$_grpcCounts (paid attempt retained: call finalize again',
      'a truncated not-retained hint':
          '$_grpcCounts (stored chunks persist; re-prepare the same con',
    };
    unreadable.forEach((label, message) {
      test('gRPC message with $label stops at once to reconcile', () async {
        final client = _GrpcMessageClient(message);
        addTearDown(client.close);
        final logs = <String>[];
        final backoffs = <int>[];
        Object? error;
        try {
          await finalizeWithRetry(
            client,
            'up-1',
            {'0xq': '0xt'},
            backoff: (attempt) {
              backoffs.add(attempt);
              return Duration.zero;
            },
            log: logs.add,
          );
        } catch (e) {
          error = e;
        }
        expect(error,
            _partialError(failed: 12, retryable: false, retentionKnown: false));
        expect(client.calls, equals(1));
        expect(backoffs, isEmpty);
        expect(logs.single,
            allOf(contains('retention is unknown'), contains('reconcile')));
      });
    });

    test('gRPC message with the not-retained hint is rethrown as known',
        () async {
      // Only the daemon's explicit not-retained hint means "nothing
      // retained": the error reaches the caller as known, not retryable,
      // and the helper does not log the unknown-retention advice.
      final client = _GrpcMessageClient('$_grpcCounts (stored chunks '
          'persist; re-prepare the same content to retry only the remainder)');
      addTearDown(client.close);
      final logs = <String>[];
      Object? error;
      try {
        await finalizeWithRetry(client, 'up-1', {'0xq': '0xt'},
            backoff: (_) => Duration.zero, log: logs.add);
      } catch (e) {
        error = e;
      }
      expect(error, _partialError(failed: 12, retryable: false));
      expect(client.calls, equals(1));
      expect(logs, isEmpty);
    });

    test('other errors are rethrown after one call', () async {
      final network = _FinalizeDaemon([() => _error(502, 'unreachable')]);
      final (networkError, _) = await _thrown(network);
      expect(networkError, isA<NetworkError>());
      expect(networkError, isNot(isA<PartialUploadError>()));
      expect(network.calls, equals(1));

      final bad = _FinalizeDaemon([() => _error(400, 'unknown upload_id')]);
      final (badError, _) = await _thrown(bad);
      expect(badError, isA<BadRequestError>());
      expect(bad.calls, equals(1));
    });

    test('maxAttempts below 1 is rejected before any call', () async {
      final daemon = _FinalizeDaemon([_finalized]);
      final (error, _) = await _thrown(daemon, maxAttempts: 0);
      expect(error, isA<ArgumentError>());
      expect(daemon.calls, equals(0));
    });
  });
}

File: antd-js/examples/finalize-with-retry.test.ts
import { afterEach, describe, expect, it, vi } from "vitest";

import { NotFoundError, PartialUploadError } from "../src/errors.js";
import { RestClient } from "../src/rest-client.js";
import { finalizeWithRetry } from "./finalize-with-retry.js";

// The example's retry helper repeats a finalize only for a partial the daemon
// confirmed it retained, and every stop rethrows the original typed error so
// the caller can still tell the three recovery cases apart.

const noDelay = vi.fn(() => Promise.resolve());
const log = vi.fn();
const opts = { delay: noDelay, log, maxAttempts: 3 };

function partial(
  chunksFailed: number,
  flags: { retryable?: boolean; retentionKnown?: boolean },
): PartialUploadError {
  return new PartialUploadError(`Partial upload: ${10 - chunksFailed}/10 chunks stored`, {
    chunksStored: 10 - chunksFailed,
    chunksFailed,
    totalChunks: 10,
    ...flags,
  });
}

const retained = (chunksFailed: number) => partial(chunksFailed, { retryable: true });

/** A finalize stub that throws each error in turn, then resolves `result`. */
function finalizeStub<T>(errors: unknown[], result?: T) {
  let call = 0;
  return vi.fn(async () => {
    const err = errors[call++];
    if (err !== undefined) throw err;
    return result as T;
  });
}

/** The error a promise rejects with; fails the test if it resolves. */
function rejection(p: Promise<unknown>): Promise<unknown> {
  return p.then(
    () => {
      throw new Error("expected finalizeWithRetry to throw");
    },
    (e: unknown) => e,
  );
}

function json(status: number, body: unknown): Response {
  return new Response(JSON.stringify(body), {
    status,
    headers: { "Content-Type": "application/json" },
  });
}

afterEach(() => {
  vi.clearAllMocks();
  vi.unstubAllGlobals();
});

describe("finalizeWithRetry (examples/finalize-with-retry.ts)", () => {
  it("returns the first success without retrying", async () => {
    const attempt = finalizeStub([], "done");
    await expect(finalizeWithRetry("u1", attempt, opts)).resolves.toBe("done");
    expect(attempt).toHaveBeenCalledTimes(1);
    expect(noDelay).not.toHaveBeenCalled();
  });

  it("repeats the same call while the daemon retains the attempt and progress continues", async () => {
    const attempt = finalizeStub([retained(4), retained(2)], "done");
    await expect(finalizeWithRetry("u1", attempt, opts)).resolves.toBe("done");
    expect(attempt).toHaveBeenCalledTimes(3);
    expect(noDelay.mock.calls).toEqual([[1], [2]]);
  });

  it("rethrows the original PartialUploadError when the attempts run out", async () => {
    const last = retained(1);
    const attempt = finalizeStub([retained(3), retained(2), last], "never");
    const err = await rejection(finalizeWithRetry("u1", attempt, opts));
    expect(err).toBe(last);
    expect(attempt).toHaveBeenCalledTimes(3);
    expect(log.mock.calls.at(-1)?.[0]).toMatch(/stays retained under upload_id u1/);
  });

  it("rethrows the original PartialUploadError when chunksFailed stops shrinking", async () => {
    const stalled = retained(2);
    const attempt = finalizeStub([retained(2), stalled], "never");
    const err = await rejection(finalizeWithRetry("u1", attempt, opts));
    expect(err).toBe(stalled);
    expect(attempt).toHaveBeenCalledTimes(2);
  });

  it("stops at once on confirmed non-retention and rethrows the error", async () => {
    const notRetained = partial(2, { retentionKnown: true, retryable: false });
    const attempt = finalizeStub([notRetained], "never");
    const err = await rejection(finalizeWithRetry("u1", attempt, opts));
    expect(err).toBe(notRetained);
    expect(attempt).toHaveBeenCalledTimes(1);
    expect(noDelay).not.toHaveBeenCalled();
  });

  it("stops at once on unknown retention: no retry, error rethrown, reconcile advised", async () => {
    const unknown = partial(2, {});
    const attempt = finalizeStub([unknown], "never");
    const err = await rejection(finalizeWithRetry("u1", attempt, opts));
    expect(err).toBe(unknown);
    expect((err as PartialUploadError).retentionKnown).toBe(false);
    expect(attempt).toHaveBeenCalledTimes(1);
    expect(noDelay).not.toHaveBeenCalled();
    expect(log.mock.calls.at(-1)?.[0]).toMatch(/keep upload_id u1 .*reconcile/);
  });

  it("stops when a later attempt reports unknown retention", async () => {
    const unknown = partial(1, {});
    const attempt = finalizeStub([retained(2), unknown], "never");
    const err = await rejection(finalizeWithRetry("u1", attempt, opts));
    expect(err).toBe(unknown);
    expect(attempt).toHaveBeenCalledTimes(2);
  });

  it("passes any other error through untouched", async () => {
    const gone = new NotFoundError("upload_id u1 not found");
    const attempt = finalizeStub([retained(2), gone], "never");
    const err = await rejection(finalizeWithRetry("u1", attempt, opts));
    expect(err).toBe(gone);
    expect(attempt).toHaveBeenCalledTimes(2);
  });

  describe("through RestClient.finalizeUpload", () => {
    const partialBody = (extra: Record<string, unknown>) => ({
      error: "Partial upload: 10/12 chunks stored, 2 failed after retries: quorum",
      code: "PARTIAL_UPLOAD",
      chunks_stored: 10,
      chunks_failed: 2,
      total_chunks: 12,
      ...extra,
    });

    it("resends the same upload_id and tx_hashes until the retained attempt completes", async () => {
      const fetchFn = vi
        .fn()
        .mockResolvedValueOnce(json(502, partialBody({ retryable: true })))
        .mockResolvedValueOnce(
          json(200, { chunks_stored: 12, data_map: "dm", data_map_address: "dma" }),
        );
      vi.stubGlobal("fetch", fetchFn);
      const client = new RestClient({ baseUrl: "http://localhost:8082" });
      const txHashes = { "0xq1": "0xtx1" };

      const fin = await finalizeWithRetry(
        "up-1",
        () => client.finalizeUpload("up-1", txHashes),
        opts,
      );

      expect(fin.chunksStored).toBe(12);
      expect(fetchFn).toHaveBeenCalledTimes(2);
      const bodies = fetchFn.mock.calls.map(([, init]) =>
        JSON.parse((init as RequestInit).body as string),
      );
      expect(bodies).toEqual([
        { upload_id: "up-1", tx_hashes: txHashes },
        { upload_id: "up-1", tx_hashes: txHashes },
      ]);
    });

    it("stops on an older daemon's partial (no retryable) and rethrows the typed error", async () => {
      const fetchFn = vi.fn().mockResolvedValue(json(502, partialBody({})));
      vi.stubGlobal("fetch", fetchFn);
      const client = new RestClient({ baseUrl: "http://localhost:8082" });

      const err = await rejection(
        finalizeWithRetry("up-1", () => client.finalizeUpload("up-1", {}), opts),
      );

      expect(err).toBeInstanceOf(PartialUploadError);
      expect(err).toMatchObject({ retryable: false, retentionKnown: false, chunksFailed: 2 });
      expect(fetchFn).toHaveBeenCalledTimes(1);
    });
  });
});
